polygeist/tools/cgeist/Test/Verification/sycl/issue7835.cpp

```diff
@@ -0,0 +1,60 @@
+// CHECK-MLIR: %18 = "polygeist.pointer2memref"(%17) : (!llvm.ptr<!llvm.struct<(!sycl_range_1_, !llvm.struct<(memref<?xi32, 4>)>)>, 4>) -> memref<?x!llvm.struct<(!sycl_range_1_, !llvm.struct<(memref<?xi32, 4>)>)>, 4>
+// CHECK-MLIR: affine.store %15, %alloca[0] : memref<1x!sycl_item_1_>
+// CHECK-MLIR: sycl.call(%18, %cast) {FunctionName = @"operator()", MangledFunctionName = @_ZNK4sycl3_V16detail18RoundedRangeKernelINS0_4itemILi1ELb1EEELi1EZ4testRNS0_5queueEEUlNS0_2idILi1EEEE_EclES4_, TypeName = @RoundedRangeKernel} : (memref<?x!llvm.struct<(!sycl_range_1_, !llvm.struct<(memref<?xi32, 4>)>)>, 4>, memref<?x!sycl_item_1_>) -> ()
+// CHECK-MLIR: gpu.return
+
+// CHECK-LLVM-LABEL: define weak_odr spir_kernel void @_ZTSN4sycl3_V16detail18RoundedRangeKernelINS0_4itemILi1ELb1EEELi1EZ4testRNS0_5queueEEUlNS0_2idILi1EEEE_EE
+// CHECK-LLVM-SAME:     (%"class.sycl::_V1::range.1"* noundef byval(%"class.sycl::_V1::range.1") align 8 %0, 
+// CHECK-LLVM-SAME:     { { i32 addrspace(1)*, i32 addrspace(1)*, i64, [1 x i64], [1 x i64] } }* noundef byval({ { i32 addrspace(1)*, i32 addrspace(1)*, i64, [1 x i64], [1 x i64] } }) align 8 %1) local_unnamed_addr #0 {
+// CHECK-LLVM-NEXT:  %3 = addrspacecast %"class.sycl::_V1::range.1"* %0 to %"class.sycl::_V1::range.1" addrspace(4)*
+// CHECK-LLVM-NEXT:  %4 = getelementptr %"class.sycl::_V1::range.1", %"class.sycl::_V1::range.1" addrspace(4)* %3, i64 0, i32 0, i32 0, i64 0
+// CHECK-LLVM-NEXT:  %5 = load i64, i64 addrspace(4)* %4, align 8
+// CHECK-LLVM-NEXT:  %.elt1.i = getelementptr inbounds { { i32 addrspace(1)*, i32 addrspace(1)*, i64, [1 x i64], [1 x i64] } }, { { i32 addrspace(1)*, i32 addrspace(1)*, i64, [1 x i64], [1 x i64] } }* %1, i64 0, i32 0, i32 1
+// CHECK-LLVM-NEXT:  %6 = bitcast i32 addrspace(1)** %.elt1.i to i32 addrspace(4)**
+// CHECK-LLVM-NEXT:  %7 = addrspacecast i32 addrspace(4)** %6 to i32 addrspace(4)* addrspace(4)*
+// CHECK-LLVM-NEXT:  %.unpack2.i = load i32 addrspace(4)*, i32 addrspace(4)* addrspace(4)* %7, align 8
+// CHECK-LLVM-NEXT:  %8 = load <3 x i64>, <3 x i64> addrspace(4)* addrspacecast (<3 x i64> addrspace(1)* @__spirv_BuiltInGlobalInvocationId to <3 x i64> addrspace(4)*), align 32
+// CHECK-LLVM-NEXT:  %9 = extractelement <3 x i64> %8, i64 0
+// CHECK-LLVM-NEXT:  %10 = icmp slt i64 %9, 2147483648
+// CHECK-LLVM-NEXT:  tail call void @llvm.assume(i1 %10)
+// CHECK-LLVM-NEXT:  %.not = icmp slt i64 %9, %5
+// CHECK-LLVM-NEXT:  br i1 %.not, label %11, label %14
+// CHECK-LLVM:       11: 
+// CHECK-LLVM-NEXT:    %12 = trunc i64 %9 to i32
+// CHECK-LLVM-NEXT:    %13 = getelementptr i32, i32 addrspace(4)* %.unpack2.i, i64 %9
+// CHECK-LLVM-NEXT:    store i32 %12, i32 addrspace(4)* %13, align 4
+// CHECK-LLVM-NEXT:    br label %14
+// CHECK-LLVM:       14:
+// CHECK-LLVM-NEXT:    ret void
+
+int test(sycl::queue &q) {
+  int *x = sycl::malloc_device<int>(10, q);
+  q.parallel_for(sycl::range(10), [=](sycl::id<1> id) { x[id] = id; });
```

Comment:
In what situation `ReturnVal` is null? Wondering what is special in this test case to trigger the bug.

Reply:
This happens when a cast expr is contained in a function that does not have a return stmt.  For the example we have:
```
CompoundStmt 0x7f592345b9f8
|-DeclStmt 0x7f5923458290
| `-VarDecl 0x7f59234581e0  used __SYCLKernel 'class (lambda at issue7835.cpp:8:31)' cinit
|   `-InitListExpr 0x7f5923458248 'class (lambda at issue7835.cpp:8:31)'
|     `-ImplicitCastExpr 0x7f5923458390 'int *' <AddressSpaceConversion>
|       `-ImplicitCastExpr 0x7f5923458378 '__global int *' <LValueToRValue>
|         `-DeclRefExpr 0x7f5923458358 '__global int *' lvalue ParmVar 0x7f59234582f0 '_arg_x' '__global int *'
`-CompoundStmt 0x7f592345b9e0
  `-ExprWithCleanups 0x7f592345b9c8 'void':'void'
    `-CXXOperatorCallExpr 0x7f592345b990 'void':'void' '()'
      |-ImplicitCastExpr 0x7f5923459eb0 'void (*)(sycl::id<1>) const' <FunctionToPointerDecay>
      | `-DeclRefExpr 0x7f5923459e90 'void (sycl::id<1>) const' lvalue CXXMethod 0x7f59241f8d60 'operator()' 'void (sycl::id<1>) const'
      |-ImplicitCastExpr 0x7f5923459ec8 'const class (lambda at issue7835.cpp:8:31)' lvalue <NoOp>
      | `-DeclRefExpr 0x7f59234583a8 'class (lambda at issue7835.cpp:8:31)' lvalue Var 0x7f59234581e0 '__SYCLKernel' 'class (lambda at issue7835.cpp:8:31)'
      `-ImplicitCastExpr 0x7f592345b978 'sycl::id<1>':'class sycl::id<1>' <ConstructorConversion>
        `-CXXConstructExpr 0x7f592345b948 'sycl::id<1>':'class sycl::id<1>' 'void (ParamTy<1, 1, const item<1, true> > &)'
          `-MaterializeTemporaryExpr 0x7f592345b930 'const decltype(getItem<1, true>())':'const class sycl::item<1, true>' lvalue
            `-CallExpr 0x7f5923458418 'decltype(getItem<1, true>())':'const class sycl::item<1, true>'
              |-ImplicitCastExpr 0x7f5923458400 'auto (*)(item<1, true> *) -> decltype(getItem<1, true>())' <FunctionToPointerDecay>
              | `-DeclRefExpr 0x7f5923450fc8 'auto (item<1, true> *) -> decltype(getItem<1, true>())' lvalue CXXMethod 0x7f5923450ea8 'getElement' 'auto (item<1, true> *) -> decltype(getItem<1, true>())' (FunctionTemplate 0x3df7fd0 'getElement')
              `-CallExpr 0x7f59234583e0 'class sycl::item<1, true> *'
                `-ImplicitCastExpr 0x7f59234583c8 'class sycl::item<1, true> *(*)(void)' <FunctionToPointerDecay>
                  `-DeclRefExpr 0x7f592344f568 'class sycl::item<1, true> *(void)' lvalue Function 0x7f592344f468 'declptr' 'class sycl::item<1, true> *(void)' (FunctionTemplate 0x3df0640 'declptr')
Expr: ImplicitCastExpr 0x7f5923458390 'int *' <AddressSpaceConversion>
`-ImplicitCastExpr 0x7f5923458378 '__global int *' <LValueToRValue>
  `-DeclRefExpr 0x7f5923458358 '__global int *' lvalue ParmVar 0x7f59234582f0 '_arg_x' '__global int *'
```
where we can see that `ImplicitCastExpr 0x7f5923458390` is in a function that does not return.

Conversely if variable `x` in the code at line 57 was:
```
 accessor<int, 1, access::mode::write, access::target::global_buffer> x;
```
the AST would look like:
```
CompoundStmt 0x7f5cdda7bad8
|-DeclStmt 0x7f5cdda7b9e0
| `-VarDecl 0x7f5cdda79bc0  used LinearIndex 'const size_t':'const unsigned long' cinit
|   `-CXXMemberCallExpr 0x7f5cdda7a078 'size_t':'unsigned long'
|     |-MemberExpr 0x7f5cdda7a030 '<bound member function type>' ->getLinearIndex 0x7f5cdda79f18
|     | `-CXXThisExpr 0x7f5cdda7a020 'const class sycl::accessor<int, 1, sycl::access::mode::write, sycl::access::target::global_buffer, sycl::access::placeholder::false_t> *' implicit this
|     `-CXXConstructExpr 0x7f5cdda7b958 'id<1>':'class sycl::id<1>' 'void (const class sycl::id<1> &) noexcept'
|       `-ImplicitCastExpr 0x7f5cdda7b910 'const class sycl::id<1>' lvalue <NoOp>
|         `-DeclRefExpr 0x7f5cdda79ca0 'id<1>':'class sycl::id<1>' lvalue ParmVar 0x7f5cde772950 'Index' 'id<1>':'class sycl::id<1>'
`-ReturnStmt 0x7f5cdda7bac8
  `-ImplicitCastExpr 0x7f5cdda7bab0 'detail::const_if_const_AS<AS, int>':'int' lvalue <AddressSpaceConversion>
    `-ArraySubscriptExpr 0x7f5cdda7ba90 'typename detail::DecoratedType<int, AS>::type':'__global int' lvalue
      |-CXXMemberCallExpr 0x7f5cdda7ba38 'ConcreteASPtrType':'typename detail::DecoratedType<int, AS>::type *'
      | `-MemberExpr 0x7f5cdda7ba08 '<bound member function type>' ->getQualifiedPtr 0x7f5cde83a128
      |   `-CXXThisExpr 0x7f5cdda7b9f8 'const class sycl::accessor<int, 1, sycl::access::mode::write, sycl::access::target::global_buffer, sycl::access::placeholder::false_t> *' implicit this
      `-ImplicitCastExpr 0x7f5cdda7ba78 'size_t':'unsigned long' <LValueToRValue>
        `-DeclRefExpr 0x7f5cdda7ba58 'const size_t':'const unsigned long' lvalue Var 0x7f5cdda79bc0 'LinearIndex' 'const size_t':'const unsigned long'
```

Here we can see that `ImplicitCastExpr 0x7f5cdda7bab0` is rooted by a `ReturnStmt`. Code in clang-mlir.cc initialize `ReturnVal` only when a function contains a return stmt.

In general the cast expr may show up in functions that do not return, so it doesn't make any sense to rely in `ReturnVal` to be initialized.